### PR TITLE
#166074599 Set up the jenkins CI/CD pipeline

### PR DESCRIPTION
**Pull request by SilasKenneth**
## What does this PR do?
- Creates Jenkinsfile and writes pipeline to test, build, Codeclimate report and deploy to Kubernetes.
- Implement slack notifications at each stage of the pipeline.

## Description of Task to be completed?
**Why**

**As Jenkins**
I want to work as the CI server for Converge API product so that the CI/CD jobs handled by CircleCI would be handled by me.

**Acceptance Criteria**

```gherkin
Scenario: 
Given that an engineer implements a task
When the branch is pushed
Then Jenkins should trigger a job to run the CI/CD pipeline as specified in the Jenkinsfile
```

## How should this be manually tested?

Rerun the most recent build from [here](https://jenkins.andela.com/blue/organizations/jenkins/mrm_api/activity)

**Any background context you want to provide?**

- Circle CI config file now contains ```type: approval``` so that approval would be required before CircleCI could be used for CI/CD

**What are the relevant pivotal tracker stories?**

[#166074599](https://www.pivotaltracker.com/story/show/166074599)

**Screenshots (if appropriate)**

Screenshot 2019-05-29 18 58 13

Screenshot 2019-06-07 11 29 19
Screenshot 2019-06-07 11 31 49

**Questions:**

None

### CODE REVIEW

File: jenkins/config.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
+  apiVersion: v1
+  kind: Pod
+  metadata:
+      name: dind
```
Comment (HarithJ) [inline jenkins/config.yml]: Please use a descriptive name for your pod, it would be useful when debugging.

File: jenkins/scripts/build_image.sh [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  gsutil cp gs://${BACKEND_BASE_IMAGE_VERSION_PATH_SANDBOX}current_version .
+  VERSION=$(cat current_version)
+  export IMAGE=$BACKEND_BASE_IMAGE_SANDBOX:$VERSION
+  sed -i "s|{{IMAGE}}|$IMAGE|g" docker/prod/Dockerfile
+  docker build -f docker/prod/Dockerfile -t ${SANDBOX_BACKEND_IMAGE}:$GIT_HASH .
+  docker push ${SANDBOX_BACKEND_IMAGE}:$GIT_HASH
+  touch current_version
+  echo ${GIT_HASH} > current_version
+  gsutil cp current_version gs://${SANDBOX_BACKEND_IMAGE_VERSION_PATH}
+fi
```
Comment (HarithJ) [inline jenkins/scripts/build_image.sh]: This file is a mess! Please make use of functions and comments to make this file human readable!

File: jenkins/scripts/deploy_backend_to_kubernetes.sh [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+terraform_values() {
+    #Generate values for terraform variables
+    cd ${HOME}/deployments/
+    mkdir -p ${HOME}/deployments/secrets
+    echo $CERTIFICATE | base64 -d > ${HOME}/deployments/secrets/ssl_andela_certificate.crt
+    echo $KEY | base64 -d > ${HOME}/deployments/secrets/ssl_andela_key.key
+    # Substitute environment variable names in the deployment scripts that are
+    # specific to CIRCLECI to match Jenkins
+    sed -i -- 's/base64 --decode/base64 -d/g' ${HOME}/deployments/.circleci/deploy_to_kubernetes.sh
```
Comment (HarithJ) [inline jenkins/scripts/deploy_backend_to_kubernetes.sh]: circleci???

File: jenkins/scripts/notify_slack.sh [hunk 4]
```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env bash
+
+# get build status and branch
+# send notification
+
+STATUS=$1 #job status
```
Comment (HarithJ) [inline jenkins/scripts/notify_slack.sh]: Instead of using doing status check like this, you can make use of `post` function in Jenkinsfile.

**Review by HarithJ — CHANGES_REQUESTED** [review]

File: jenkins/scripts/notify_slack.sh (hunk 4 above)
Comment (SilasKenneth) [reply jenkins/scripts/notify_slack.sh]: Can you please expound on this @HarithJ with reference to the `Jenkinsfile`.

Comment (SilasKenneth) [reply jenkins/scripts/notify_slack.sh]: I am using one script to notify Slack of the build result per stage, the script requires as the first argument, the status of the job which I pass from the Jenkins at the post step.

**Review by HarithJ — APPROVED** [review]

**Review by mariamiah — APPROVED** [review]
LGTM

**Pull request closed without merging**